packages/aws-cdk/lib/toolkit/cli-io-host.ts

```diff
@@ -41,36 +55,87 @@ interface IoMessage {
 
 export type IoMessageLevel = 'error' | 'warn' | 'info' | 'debug' | 'trace';
 
-export type IoAction = 'synth' | 'list' | 'deploy' | 'destroy';
+/**
+ * The current action being performed by the CLI. 'none' represents the absence of an action.
+ */
+export type IoAction = 'synth' | 'list' | 'deploy' | 'destroy' | 'none';
 
 /**
- * Options for the CLI IO host.
+ * A simple IO host for the CLI that writes messages to the console.
  */
-interface CliIoHostOptions {
+export class CliIoHost {
   /**
-   * If true, the host will use TTY features like color.
+   * Returns the singleton instance
    */
-  useTTY?: boolean;
+  static getIoHost(): CliIoHost {
+    if (!CliIoHost.instance) {
+      CliIoHost.instance = new CliIoHost();
+    }
+    return CliIoHost.instance;
+  }
 
   /**
-   * Flag representing whether the current process is running in a CI environment.
-   * If true, the host will write all messages to stdout, unless log level is 'error'.
-   *
-   * @default false
+   * Determines which output stream to use based on log level and configuration.
    */
-  ci?: boolean;
-}
+  public static getStream(level: IoMessageLevel, forceStdout: boolean) {
+    // For legacy purposes all log streams are written to stderr by default, unless
+    // specified otherwise, by passing `forceStdout`, which is used by the `data()` logging function, or
+    // if the CDK is running in a CI environment. This is because some CI environments will immediately
+    // fail if stderr is written to. In these cases, we detect if we are in a CI environment and
+    // write all messages to stdout instead.
+    if (forceStdout) {
+      return process.stdout;
+    }
+    if (level == 'error') return process.stderr;
+    return this.ci ? process.stdout : process.stderr;
+  }
 
-/**
- * A simple IO host for the CLI that writes messages to the console.
- */
-export class CliIoHost {
-  private readonly pretty_messages: boolean;
-  private readonly ci: boolean;
+  /**
+   * Singleton instance of the CliIoHost
+   */
+  private static instance: CliIoHost | undefined;
+
+  /**
+   * Whether the host should apply chalk styles to messages. Defaults to false if the host is not running in a TTY.
+   */
+  private isTTY: boolean;
+
+  /**
+   * Whether the CliIoHost is running in CI mode. In CI mode, all non-error output goes to stdout instead of stderr.
+   */
+  private ci: boolean = false;
```

Comment:
i'm not against this style, but i would like how you add defaults to `ci` and `isTTY` to be the same...